kelpie-stream10-dev-BJS2-77395

JegorKuz opened the pull request.

JegorKuz committed:
[UPDATE]
Решение для таски BJS2-77219.
Нет тестов.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Исправил ошибки по оформлению и логике.
Переделал фильтр в стратегию.
Заменил кастомный SQL "create" запрос на существующий "save".
Заменил реализации сервисом на интерфейсы.
Изменил выброс пустой ошибки, на ошибку с сообщением.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Избавился от Optional изменив логику.
Дополнил решение для метода create - созданием скилов для entity.
Убрал не нужный SkillRequestService/Impl.
Добавил более подробные сообщения для Exceptions.
Убрал не нужные проверки в Controller.
Убрал не нужный метод из RecommendationRequestRepository.

[TODO]
Исправить фильтр.
До делать тесты.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Исправио фильтр.

[TODO]
До делать тесты.

Commit:
add: Карьера BJS2-77378

littlerock2877 committed:
add: unit tests

Commit:
fix: Карьера BJS2-77378

littlerock2877 committed:
Merge branch 'kelpie-master-stream10' into develop/BJS2-77387

littlerock2877 committed:
fix: compillation

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Убрал не нужный чек для create.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77395.
Сделал тесты кроме фильтров.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77395.
Сделал тесты кроме фильтров.

Commit:
fix2: Карьера BJS2-77378

AlexandrSid committed:
Приведение командного мейна в рабочее состояние

littlerock2877 committed:
fix: tests

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77395.
Сделал тесты.

[TODO]
Протестировать отдельно фильтры.

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77395.
Сделал тесты.

[TODO]
Протестировать отдельно фильтры.

miked4king committed:
Merge pull request #3013 from CorporationX/bugfix/master-workstate-fix

Приведение командного мейна в рабочее состояние

littlerock2877 committed:
Merge remote-tracking branch 'refs/remotes/origin/kelpie-master-stream10' into develop/BJS2-77387

# Conflicts:
#	src/main/java/school/faang/user_service/dto/UserDto.java

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77219.
Исправил ошибки по комментариям.

miked4king committed:
Merge pull request #2973 from CorporationX/kelpie-stream10-dev-BJS2-77219

Kelpie stream10 dev bjs2 77219

AdminSyom committed:
Merge branch 'kelpie-master-stream10' into develop/BJS2-77378

Commit:
fix3: Карьера BJS2-77378

AlexandrSid committed:
77371
ребейз на мастер и приведение в готовое к мержу состояние

Commit:
fix4: Карьера BJS2-77378

AlexandrSid committed:
77371
refactoring

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+    @Spy
+    private RecommendationRequestMapperImpl recommendationRequestMapper;
+
+    private RecommendationRequestServiceImpl recommendationRequestService;
```
над сервисом должна быть  @InjectMocks. У тебя разве без неё тесты проходят?

Bobritta replied:
наверное проходят из-за конструктора ниже. Но зачем?

JegorKuz replied:
Всё верно, из-за конструктора ниже. Он же нужен, для того чтобы замокать фильтры, как сделать фильтры через мок(если это вообще возможно) без конструктора в лекции не показали, сам искать даже не пробовал.

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+
+    @Test
+    public void testCreateThrowsRequesterIsNotFound() {
+        long requesterId = 1L;
```
Всем советуют вот такие повторяющиеся вводные выносить в сетап в @BeforeEach. Просто, чтобы не путаться, м.б. использовать разные значения?
        long requesterId = 1L;
        long receiverId = 2L;
        long recommendationRequestId1 = 133L;
        long recommendationRequestId2 = 233L;
        , например?

JegorKuz replied:
Правильно подметила.

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+    public void testCreateThrowsRequesterIsNotFound() {
+        long requesterId = 1L;
+
+        RecommendationRequestDto dto = new RecommendationRequestDto();
```
тоже повторяется строка, тоже бы вынести

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(userRepository.findById(requesterId))
+                .thenReturn(Optional.empty());
+
+        assertThrows(IllegalArgumentException.class, () -> recommendationRequestService.create(dto));
```
1. у тебя же метод в таком случае выкидывает EntityNotFoundException, а не IllegalArgumentException.
2.  Можно будет результат метода присвоить:
EntityNotFoundException e = assertThrows(IllegalArgumentException.class, () -> recommendationRequestService.create(dto));
чтобы потом через асертИквалс сравнить сообщения ожидаемое и e.getMessage

Думаю, это надо сделать потому, что такой же класс исключения, но с сообщением о ненайденном ресивере тоже може выброситься здесь

JegorKuz replied:
1. Правильно, надо исправить.
2. Добавить можно, как доп меру безопасности на случай если кто-то изменит код. В данном сценарии не может выброситься ресивер, тк возвращаем эмпти, первый запрос findById выбросит ошибку и на этом тесты закончатся.

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(userRepository.findById(requesterId))
+                .thenReturn(Optional.of(new User()));
+        when(userRepository.findById(receiverId))
+                .thenReturn(Optional.empty());
```
Кажется, здесь when(userRepository.findById(receiverId)) можно закончить thenThrow ?

JegorKuz replied:
Если мы используем thenThrow, то мы форсим код вернуть ошибку которую мы захардкодим в thenThrow. Тогда теряется смысл проверки, тк мы проверяем что программа и сама выкинет ошибку, без нашего вмешательства.

Bobritta replied:
согласна, поняла

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(userRepository.findById(receiverId))
+                .thenReturn(Optional.empty());
+
+        assertThrows(IllegalArgumentException.class, () -> recommendationRequestService.create(dto));
```
Тоже тип исключения и вопрос о том, кого именно не нашли в базе: реквестера или ревьюера

JegorKuz replied:
Правильно.

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(recommendationRequestRepository.findLatestPendingRequest(requesterId, receiverId))
+                .thenReturn(Optional.of(entity));
+
+        assertThrows(IllegalArgumentException.class, () -> recommendationRequestService.create(dto));
```
Вроде здесь на логику не влияет, но странно, что в итоге ты вызываешь метод, передавая туда дто с проставленным апдейтедЭт датой месяц назад. Как будто по идее твой реальный метод должен будет не с такими данными вызываться.
М.б. сначала дто превратить в энтити, потом у энтити проставить апдейтедЭт? Будет, вроде, логичнее.

Bobritta replied:
NullPointerException, кстати, здесь ещё поймать можно (51 строка сервиса):
 if (recommendationRequest.getUpdatedAt().isAfter(LocalDateTime.now().minus(Period.ofMonths(6)))) {
            throw new IllegalArgumentException("Recommendation request has already been updated in the last 6 months.");
        }

Bobritta replied:
реквесты в методе криейт же не проверяются на заполненность этого поля, так что запросто в базе окажутся с нулевым апдейтом, так?

JegorKuz replied:
Про NPE всё верно, проверки на это нет, предполагается что дата присутствиует, если запрос не был создан до этого и мы добавляем новый, то ошибка выскочит 100%.

С первым не совсем согласен, если буду править код, то дата будет добавляться уже в сущность(RecommendationRequest), но только при условии, что она не передалась из дто. В тестах предполагается что она мне уже дана в дто и поэтому здесь идёт в том же порядке.

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        RecommendationRequestDto dto = new RecommendationRequestDto();
+        dto.setRequesterId(requesterId);
+        dto.setReceiverId(receiverId);
+        dto.setUpdatedAt(date);
```
Я думаю, что поле UpdatedAt в ДТО дано, чтобы возвращаемы данные из базы репозиторием давали ответ, почему новый запрос делать ещё слишком рано. Но странно, что ты передаешь в запросе это поле заполненным

JegorKuz replied:
Не понял?
База вернёт ответ и без поля UpdatedAt. 

Что значит "Почему новый запрос делать ещё слишком рано"?

Bobritta replied:
По идее, юзер1 просит юзера2 о рекомендации. И только твоя логика скажет, может он это сделать или слишком мало времени прошло от прошлого запроса.
А у тебя получается, что юзер1 приходит и говорит: дай мне рекомендацию на такие-то скилы, обновлнную такого-то числа. Или как можно интерпретировать поле updatedAt в дто запроса?

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(skillRepository.existsById(1L))
+                .thenReturn(false);
+
+        assertThrows(NoSuchElementException.class, () -> recommendationRequestService.create(dto));
```
и здесь другой тип исключения и можно проверить сообщение:
EntityNotFoundException("Not all required skills exist in data base")

JegorKuz replied:
Согласен.

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(recommendationRequestRepository.save(entity))
+                .thenReturn(entity);
+
+        assertEquals(recommendationRequestMapper.toDto(entity), recommendationRequestService.create(dto));
```
Если всё же метод create при создании из ДТО с датой создания будет делать сущность, где дате апдейта будет присвоено значение даты создания, а в твоей исходной ДТО для метода этого значения не будет, то ассерт не пройдёт.
Но прямо сейчас у тебя в задаче нет такой логики. Прямо сейчас в твоей ДТО есть дата создания и дата обновления. При этом дата создания, можно предположить проставляется из LocalDateTime.now или что-то вроде того. Но что происходит с updatedAt полем ? У тебя по нему валидация. Но в какой момент оно вообще используется?

Bobritta replied:
Ммм, а вообще у тебя идёт проверка, что реквест за последние полгода был хоть какой-то от одного пользователя другому. А вдруг они же хотят устроить рекомендацию по другим скилам. скажем сначала "булочки печь", потом "трактор водить". а ты ему отказ? Наверное, тут я уже усложняю, но мысль, вроде, бизнесу была бы полезная :)

JegorKuz replied:
Со второй частью согласен, но не мне судить логику менторов, видимо пока что по этому поводу париться не приходится. Может потом добавлю по какому-то параметру проверку, если того потребует задание.

Первый вопрос перечитал раз 6 уже, всё равно не понимаю, может попробуй переформулировать, или можно в ПП, там уже развернёшь мысль по полной.

Bobritta replied:
Это в продолжение идеи о том, что создавать запрос, передавая в ДТО дату апдейта - странно. Даже дату создания - странно.
Юзер1 просит юзера2 дать рекомендацию, просит через 3 недели, апдейтед 3 недели назад? Я думаю, эти поля в ДТО только для ответа

Bobritta reviewed (commented):
до фильтров только дочитала

littlerock2877 committed:
fix: pr comments

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+
+        List<RecommendationRequestDto> listOfExpectedDtos = List.of(dto1, dto2);
+
+        RequestFilterDto requestFilterDto = new RequestFilterDto();
```
Опять же комментарий не к тестам, но к задаче. RequestFilterDto были заточены не под запрос на рекомендации, а под что-то другое. В RequestFilterDto можно передать описание, но нельзя скил или список скилов, а фильтры по ним как будто были бы полезны бизнесу. Через anyMatch и ignorCase можно было бы сравнить слова из описания со списком скиллов, но названия скилов - не обязательно одно слово.
В общем, не думаю, что стоит так усложнять задачу. Я, наверное, предложила бы использовать RecommendationRequestDto в качестве аргумента, там есть список скиллов. Интересно, почему в основной задаче решили иначе?

Bobritta replied:
Короче, это в том числе и для меня показательно: когда смотришь выполнение задачи на предмет соответствия детальному ТЗ, забываешь проверить на соответствие логике и здравому смыслу :)

JegorKuz replied:
Касательно первого, в RequestFilterDto есть skillId.
Не по RecommendationRequestDto, а по RequestFilterDto - потому что он меньше и в нём не все параметры. Как раз таки этого мы и хотим - ограничить кол-во параметров по максимуму, оставить только необходимые.

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        entity1.setReceiver(receiver);
+        RecommendationRequest entity2 = new RecommendationRequest();
+        entity2.setId(recommendationRequestId2);
+
```
тут прям много строк повторяют вводные предыдущего теста. Если не хочешь их в бефорич выносить, можно приватным методом сделать

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+                .thenReturn(List.of(entity1, entity2));
+
+        when(requesterIdFilter.isApplicable(any())).thenReturn(false);
+        when(receiverIdFilter.isApplicable(any())).thenReturn(false);
```
Для методов, которые возвращают boolean, по умолчанию Mockito будет возвращать false. Это связано с тем, что false является значением по умолчанию для примитивного типа boolean в Java. Т.е. две строки выше можно убрать вообще, получается

JegorKuz replied:
Проверил - работает, получается можно.

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        RecommendationRequestDto dto1 = recommendationRequestMapper.toDto(entity1);
+        RecommendationRequestDto dto2 = recommendationRequestMapper.toDto(entity2);
+
+        List<RecommendationRequestDto> listOfExpectedDtos = List.of(dto1, dto2);
```
Для проверки 2 фильтров сразу я бы взяла 3 энтити изначально. Под условия первого фильтра подошлибы 1 и 2 энтити, а под условия второго 1 и 3, напрммер. Тогда результат их совместного действия привёл бы к ожидаемому списку, содержащему только 1 энтити

Bobritta replied:
Тогда и следующий метод testGetRequestsPassSeparateFiltersReturnEmpty() был бы не нужен, наверное

JegorKuz replied:
Исправлено

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+
+                    return stream.filter(recommendationRequest ->
+                            recommendationRequest.getRequester().getId().equals(requester1.getId()));
+                });
```
Тоже эта запись у тебя повторяется из раза в раз. М.б. заменить и отдельно прописать как переменную функцию, чтобы её вставлять в заготовки?

ИИ предлагает так:
Function<Stream<RecommendationRequest>, Stream<RecommendationRequest>> filterFunction = stream ->
        stream.filter(recommendationRequest ->
                recommendationRequest.getRequester().getId().equals(requester1.getId()));

when(requesterIdFilter.apply(any(), any()))
        .thenAnswer((Answer<Stream<RecommendationRequest>>) invocations ->
                filterFunction.apply(invocations.getArgument(0)));

Выглядит не короче. Так что не знаю, стоит ли изменять в эту сторону или оставить кучу дублирующегося кода

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(recommendationRequestRepository.findById(id))
+                .thenReturn(Optional.empty());
+
+        assertThrows(NoSuchElementException.class, () -> recommendationRequestService.getRequest(id));
```
тоже EntityNotFoundException вместо NoSuchElementException и сообщение можно проверить

miked4king committed:
Merge pull request #2989 from CorporationX/develop/BJS2-77378

add: Карьера BJS2-77378

miked4king committed:
Merge pull request #2997 from CorporationX/develop/BJS2-77387

add: unit tests [BJS2-77387]

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(recommendationRequestRepository.findById(id))
+                .thenReturn(Optional.of(recommendationRequest));
+
+        assertEquals(expectedDto, recommendationRequestService.getRequest(id));
```
этим тестом ты проверяешь, что равны значения, напрямую указанные тобой же:
1. recommendationRequestMapper.toDto(recommendationRequest) - проверка работы маппера, что, кажется, не имеет смысла. Можно вообще ДТО не создавать и сравнить исходный айди и тот, что будет принадлежать запросу, полученному в recommendationRequestService.getRequest(id).
2. recommendationRequestService.getRequest(id) - проверка, что вызовется метод recommendationRequestRepository.findById(id). Ну так и зачем передавать мок-объекту и проверять, возвращаемый мок-обхектом результат? для логики метода сервиса не дстаточно ли просто проверить через верифай, что recommendationRequestRepository.findById(id) вообще вызывается?

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(recommendationRequestRepository.findById(id))
+                .thenReturn(Optional.of(recommendationRequest));
+
+        assertThrows(IllegalArgumentException.class, () -> recommendationRequestService.rejectRequest(id, rejectionDto));
```
Здесь бы может и в основной логике в сообщении дополнить по какой причине "Unable to reject request" 
: "помоту что статус уже" через стринг.формат и подстановку "отклонено/принято". А то бедный юзер будет тыкать и тыкать пытаясь добиться отмены, а ты - ему отвечать "Unable to reject request"

Bobritta replied:
ну и тогда в этом тесте проверить. с правильным ли сообщением выпадает исключение

Bobritta commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,440 @@
+        when(recommendationRequestRepository.findById(id))
+                .thenReturn(Optional.empty());
+
+        assertThrows(NoSuchElementException.class, () -> recommendationRequestService.rejectRequest(id, rejectionDto));
```
Аналогично:
-> new EntityNotFoundException("Recommendation request with id %s doesn't exist".formatted(id))) у тебя в методе

littlerock2877 committed:
chg: names of event and recommendation filters

AlexandrSid committed:
77371
after PR review refactoring

AlexandrSid committed:
77371
after PR review refactoring 2

miked4king committed:
Merge pull request #3041 from CorporationX/bugfix/kelpie-main-branch

chg: fix beans name conflicts in master branch for kelpie

miked4king committed:
Merge pull request #3031 from CorporationX/develop/BJS2-77371

77371

JegorKuz committed:
Merge remote-tracking branch 'origin/kelpie-stream10-dev-BJS2-77395' into kelpie-stream10-dev-BJS2-77395

JegorKuz committed:
[UPDATE] Решение для таски BJS2-77395.
Исправил ошибки по комментариям.
При обновлении ветки добавились чужие файлы.
Моя работа:
RecommendationRequestServiceTest
RecommendationRequestServiceImpl

miked4king reviewed (approved):
Хорошая работа
